chassembly/src/main/java/org/quiltmc/chasm/lang/internal/intrinsics/ReduceFunction.java

```diff
@@ -0,0 +1,44 @@
+package org.quiltmc.chasm.lang.internal.intrinsics;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.quiltmc.chasm.lang.api.ast.ListNode;
+import org.quiltmc.chasm.lang.api.ast.MapNode;
+import org.quiltmc.chasm.lang.api.ast.Node;
+import org.quiltmc.chasm.lang.api.ast.NullNode;
+import org.quiltmc.chasm.lang.api.eval.Evaluator;
+import org.quiltmc.chasm.lang.api.eval.FunctionNode;
+import org.quiltmc.chasm.lang.api.exception.EvaluationException;
+
+public class ReduceFunction extends IntrinsicFunction {
+    @Override
+    public Node apply(Evaluator evaluator, Node arg) {
+        if (arg instanceof MapNode) {
+            Map<String, Node> args = ((MapNode) arg).getEntries();
+            Node list = args.get("list");
+            Node function = args.get("func");
+
+            if (list instanceof ListNode && function instanceof FunctionNode) {
+                Map<String, Node> funcArgs = new LinkedHashMap<>();
+                MapNode funcArgsNode = new MapNode(funcArgs);
+
+                return ((ListNode) list).getEntries().stream().reduce((first, second) -> {
+                    funcArgs.put("first", first);
+                    funcArgs.put("second", second);
+
+                    return ((FunctionNode) function).apply(evaluator, funcArgsNode);
+                }).orElse(NullNode.INSTANCE);
```

Comment:
This is returned only for empty lists, is that correct?
Should we throw instead? I don't think returning null makes sense.

Reply:
Yes, it should only be returned for empty lists (or whatever causes Java's reduction implementation to fail), and I made it just return null since it allows for some additional code can be written in chassembly to handle such a case where the list is empty... though the list size could probably be checked beforehand as well
(Third opinion?)